### DISCUSSION TITLE
Allow instantiation of a testing.User without using the plugin

### DISCUSSION THREAD

**Discussion by marcuslimdw** [zauberzeug/nicegui#3508] [Ideas / Feature Requests]
In `nicegui/testing/plugin.py`, a `pytest` plugin is defined that creates `User` and `Screen` fixtures in the background available under the names `user` and `screen` respectively.

However, `user` is quite a common identifier, and it may clash with existing code that a NiceGUI consumer has. Exacerbating the problem is `pytest`'s requirement that plugins be placed in the root `conftest.py`, increasing the blast radius.

My suggestion: expose a function that *creates* a `testing.User` separately, and allow NiceGUI consumers to import just this function and use it to create their own test fixtures (along with appropriate setup of their NiceGUI applications). Retain the plugin, but simply define the fixture in terms of this function. This achieves two objectives:

1. Users who want easy setup of test fixtures can just use the plugin
2. Users who want a bit more customisation can import the appropriate functions and create their own fixtures

**Comment by rodja** [comment]
Thanks for your interest in our user simulation test framework. We tried to make the fixture as simple as possible without causing conflicts with other pytest plugins. Thats why the plugin must be "manually" activated via `pytest_plugins = ['nicegui.testing.plugin']
` instead of registering it for auto-load via pytest entrypoint. But of course besides pytest plugins there may be other fixtures called "user". Is that the case on your end? Why do you already have such a fixture in your UI-tests?

**Reply by marcuslimdw** [reply]
great question - due to...reasons...our testing strategy is to write a few E2E tests that basically mimic the whole user flow and both check that the UI is responding as expected and that the appropriate data persistence level actions are being taken. therefore, my `user` fixture populates the DB with a domain-specific `User` object.

**Reply by rodja** [reply]
Hmmm... could you create a [minimum reproducible code example](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)? I wonder if it would be possible to import your domain-specific User as a different name in the conftest.py...
The problem with making the `user` creation function available without being a fixture is that it requires the `request` fixture.

**Reply by marcuslimdw** [reply]
Either in the same test file, or in a parent `conftest.py`:
```py
import pytest


@pytest.fixture
def user():
    pass
```

And, in my test file itself:
```py
from nicegui import testing


def test_this(user):
    assert isinstance(user, testing.User)
```

This test fails as is because the local `user` fixture overrides the NiceGUI plugin's fixture. If I remove my fixture, the test passes.

**Reply by rodja** [reply]
What happens if your `user` fixture depends on the NiceGUI `user` fixture in the `conftest.py`?

```py
import pytest

pytest_plugins = ['nicegui.testing.user_plugin']

@pytest.fixture
def user(user):
    # populates the DB
    return user
```

**Reply by marcuslimdw** [reply]
This could theoretically work, but isn't really something I'd like to do, for three reasons:

1. Using a plugin means that (especially assuming I want to use type hinting) my root `conftest.py` depends on my UI layer (NiceGUI). I'd prefer to keep all my UI-related tests, *including setup logic*, in their own subfolder.
2. This fixture is used not just by my integration tests that involve NiceGUI, but also other code, which means *their* tests will also depend on NiceGUI.
3. I also need to return my actual domain user from this fixture, which means that it'd be more like `return db_user, nicegui_user`. Every consuming test that doesn't involve the UI layer will need to drop `nicegui_user` from this `tuple`, which is unwieldy.

**Reply by marcuslimdw** [reply]
@rodja, would you accept a PR that:
* extracts the "setup/create user" behaviour into a separate function that may be imported from `nicegui.testing`, and
* modifies the `user` fixture to use this new function?

**Reply by rodja** [reply]
Sounds intriguing. Go ahead!
